### PR TITLE
[JUJU-1392] Deprecation check for bootstrap/deploy/add-machine/add-model commands

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR adds series validation for:
1. add-machine with --series;
2. bootstrap with --bootstrap-series;
3. model config `default-series`;
4. juju deploy a charm/bundle;

This PR also deprecates LXD version < `5.2`.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [ ] ~Comments saying why design decisions were made~
- [ ] ~Go unit tests, with comments saying what you're testing~
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```console
$ lxc version
Client version: 5.0.0
Server version: 5.0.0

$ juju bootstrap lxd k1 --config test-mode=true --build-agent --bootstrap-series=bionic
ERROR LXD version has to be at least "5.2.0", but current version is only "5.0.0"

$ lxc version
Client version: 5.4
Server version: 5.4

$ juju bootstrap lxd k1 --config test-mode=true --build-agent --bootstrap-series=bionic
Creating Juju controller "k1" on lxd/localhost
ERROR failed to bootstrap model: use --force to override: bionic not supported

### TODO: --force for bootstrap (unknown vs deprecated series)  ###
$ juju bootstrap lxd k1 --config test-mode=true --build-agent --bootstrap-series=bionic --force
Creating Juju controller "k1" on lxd/localhost
Building local Juju agent binary version 3.0-rc1 for amd64
To configure your system to better support LXD containers, please see: https://github.com/lxc/lxd/blob/master/doc/production-setup.md
Launching controller instance(s) on localhost/localhost...

$ juju bootstrap lxd k1 --config test-mode=true --build-agent --config default-series=bionic
ERROR series "bionic" not supported

$ juju add-model default

$ juju model-config default-series
jammy

$ juju model-config default-series=bionic
ERROR series "bionic" not supported (not supported)

$ juju deploy cs:~jameinel/ubuntu-lite-6
ERROR the charm defined series "xenial, trusty" not supported

$ juju deploy cs:~jameinel/ubuntu-lite-6 --series xenial
ERROR ubuntu-lite is not available on the following series: xenial not supported

$ juju deploy cs:~jameinel/ubuntu-lite-6 --series focal
ERROR series "focal" not supported by charm, the charm supported series are: xenial,trusty. Use --force to deploy the charm anyway.

$ juju deploy cs:~jameinel/ubuntu-lite-6 --series xenial --force
ERROR ubuntu-lite is not available on the following series: xenial not supported

$ juju deploy cs:~jameinel/ubuntu-lite-6 --series focal --force
Located charm "ubuntu-lite" in charm-store, revision 6
Deploying "ubuntu-lite" from charm-store charm "ubuntu-lite", revision 6 in channel stable on focal

$ cat ./testcharms/charm-repo/bundle/multi-doc-overlays/bundle.yaml | grep series
series: trusty

$ juju deploy ./testcharms/charm-repo/bundle/multi-doc-overlays/bundle.yaml
Located charm "apache2" in charm-store, revision 26
Located charm "wordpress" in charm-store, revision 5
Executing changes:
- upload charm apache2 from charm-store for series trusty with architecture=amd64
- deploy application apache2 from charm-store on trusty
ERROR cannot deploy bundle: apache2 is not available on the following series: trusty not supported

$ juju deploy ./testcharms/charm-repo/bundle/multi-doc-overlays/bundle.yaml --force
Located charm "apache2" in charm-store, revision 26
Located charm "wordpress" in charm-store, revision 5
Executing changes:
- upload charm apache2 from charm-store for series trusty with architecture=amd64
- deploy application apache2 from charm-store on trusty
ERROR cannot deploy bundle: apache2 is not available on the following series: trusty not supported

$ juju add-machine --series=bionic
failed to create 1 machine
ERROR series "bionic" not supported

$ juju add-machine --series=focal
created machine 1

$ juju add-machine --series=jammy
created machine 2

$ juju add-machine
created machine 3

$ juju machines --format json | jq -r '.machines | to_entries | map("\(.key)=\(.value.series)") | .[]'
0=focal
1=focal
2=jammy
3=jammy

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1981955

### CODE REVIEW

**Comment by wallyworld** [comment]
A few initial comments looking at the QA steps

`ERROR failed to bootstrap model: use --force to override: bionic not supported`
we don't want to allow `--force` to override unsupported series
(is the the TODO you are referring to?)

`ERROR the charm supported series ([xenial trusty]) are deprecated not valid`
The error message structure needs fixing.
Also, xenial and trusty etc are more than just deprecated - they are not supported anymore

`ERROR series "focal" not supported by charm, supported series are: xenial,trusty. Use --force to deploy the charm anyway.`
xenial and trust are not supported; we don't want to allow `--force` at all for these

`ERROR series "bionic" not supported (not supported)`
we don't want the extra `(not supported)` at the end

**Comment by ycliuhw** [comment]
> 

So the 1st one is a TODO and will be fixed on https://bugs.launchpad.net/juju/+bug/1986595
2nd and 3rd error message issues have been fixed.
The last one is because the existing `ModelSet`  facade just simply returns an error but does not use `*params.Error`. To fix this, we will need to bump the facade version, and should be done on 2.9. I will probably fix it as a drive-by in following PRs.

File: cmd/juju/application/deployer/series_selector_test.go [hunk 1]
```diff
@@ -137,7 +137,7 @@ func (s *SeriesSelectorSuite) TestCharmSeries(c *gc.C) {
 				conf:                defaultSeries{},
 				supportedJujuSeries: set.NewStrings("bionic", "cosmic"),
 			},
-			err: "series not specified and charm does not define any",
+			err: `the charm defined series "precise" not supported`,
```
Comment (wallyworld) [inline cmd/juju/application/deployer/series_selector_test.go:140]: why this change? if the rest of the test is unchanged, shouldn't we expect the same outcome here? why does the charm define "precise" now and it didn't before?

File: cmd/juju/application/deployer/series_selector_test.go [hunk 2]
```diff
@@ -166,16 +166,6 @@ func (s *SeriesSelectorSuite) TestCharmSeries(c *gc.C) {
 			},
 			err: "series: disco not supported",
 		},
-		{
-			title: "juju deploy multiseries with force  # use model series defaults if supported by charm, force",
```
Comment (wallyworld) [inline cmd/juju/application/deployer/series_selector_test.go:170]: can we adapt this test instead of deleting it? are we losing coverage?

**Review by wallyworld — APPROVED** [review]
Some small issues with the wording of the error messages.
Nice to see "precise" etc gone

File: cmd/juju/application/deploy_test.go [hunk 3]
```diff
@@ -435,7 +436,7 @@ func (s *DeploySuite) TestDeployFromPath(c *gc.C) {
 func (s *DeploySuite) TestDeployFromPathUnsupportedSeries(c *gc.C) {
 	path := testcharms.RepoWithSeries("bionic").ClonedDirPath(c.MkDir(), "multi-series")
 	err := s.runDeploy(c, path, "--series", "quantal")
-	c.Assert(err, gc.ErrorMatches, `series "quantal" not supported by charm, supported series are: focal,jammy,bionic`)
+	c.Assert(err, gc.ErrorMatches, `multi-series is not available on the following series: quantal not supported`)
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply cmd/juju/application/deploy_test.go:439]: The existing code just prints the error instead of a slice of series, https://github.com/juju/juju/blob/4520638e6cc76701fc0362771bcdb37aa512d48c/cmd/juju/application/deployer/deployer.go#L635 I can change it.

File: cmd/juju/application/refresh_test.go [hunk 4]
```diff
@@ -374,6 +376,16 @@ func (s *RefreshErrorsStateSuite) SetUpSuite(c *gc.C) {
 		c.Skip("Mongo failures on macOS")
 	}
 	s.RepoSuite.SetUpSuite(c)
+
+	// TODO: remove this patch once we removed all the old series from tests in current package.
+	s.PatchValue(&deployer.SupportedJujuSeries,
+		func(time.Time, string, string) (set.Strings, error) {
+			return set.NewStrings(
+				"centos7", "centos8", "centos9", "genericlinux", "kubernetes", "opensuseleap",
+				"jammy", "focal", "bionic", "xenial",
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply cmd/juju/application/refresh_test.go:385]: The returned series here is different for some tests. It depends on which charm the test is using.
And this is really a very `short-term` (a few more days probably?) patch to make tests passing and can be removed once the testing charm enhancement work is done.

File: cmd/juju/application/deployer/series_selector_test.go (hunk 1 above)
Comment (ycliuhw) [reply cmd/juju/application/deployer/series_selector_test.go:140]: The old error message was not really helpful. This PR fixes it and the new error message is more helpful and also makes more sense, isn't it?

File: cmd/juju/application/deployer/series_selector_test.go (hunk 2 above)
Comment (ycliuhw) [reply cmd/juju/application/deployer/series_selector_test.go:170]: The test scenario is juju deploy with --force --series `an-unsupported-old-series`.
One of the changes of this PR is to disallow/reject this scenario.
So this test does not make sense anymore.
